docs/protobuf-libraries.md

```diff
@@ -0,0 +1,109 @@
+
+### proto3-json-serializer
+
+C
+
+- Adds spec-compliant JSON encoding to protobufjs
+- Maintained by responsive Googlers, 900k downloads/wk
+- Requires runtime-loaded messages (not compatible with generated classes)
+
+### google-protobuf
+
+B
+
+- Official Google lib, 800k downloads/wk
+- No types or JSON encoding
+- Compiler installed separately (not on npm)
+
+### ts-proto
+
+A and some of C
+
+- Generates TS interfaces and encoding functions
+- Designed for POJOs (no instances of message classes), so can't do B
+- JSON encoding is probably [not yet fully spec compliant](https://github.com/stephenh/ts-proto/pull/448#issuecomment-998166664)
+
+### protoc-gen-ts
+
+A and B
+
+- Plugin for Google's `protoc` compiler
+- Generated classes extend `google-protobuf`'s Message, but doesn't add JSON
+- Maintainer [seems interested in JSON encoding](https://github.com/protocolbuffers/protobuf/issues/4540#issuecomment-915609405), but isn't there yet (only has `to/fromObject` methods—need eg a fromJSON that converts the below base64 to a bytearray, and a toJSON that converts a bytearray to base64)
+
+## Current solution
+
+- Use `protobufjs` with `proto3-json-serializer`
+- Have users use runtime-loaded messages (not generated classes) and `Class.create` (not `new Class()`)
+- Patch `json-module` output (which adds `nested` attributes to lowercase namespaces [which causes TS error](https://github.com/protobufjs/protobuf.js/issues/1014))
+
+```ts
+// json-module.js generated with:
+// pbjs -t json-module -w commonjs -o json-module.js *.proto
+
+// protos/root.js
+const { patchRoot } = require('@temporalio/common');
+const unpatchedRoot = require('./json-module');
+module.exports = patchRoot(unpatchedRoot);
+
+// root.d.ts generated with:
+// pbjs -t static-module *.proto | pbts -o root.d.ts -
+
+// src/client.ts
+import { foo } from '../protos/root';
+
+await client.start(protoWorkflow, {
+  args: [foo.bar.ProtoInput.create({ name: 'Proto', age: 1 })],
+  taskQueue: 'tutorial',
+  workflowId: 'my-business-id',
+});
+
+// src/workflows.ts
+import { foo } from '../protos/root';
+
+export async function protoWorkflow(input: foo.bar.ProtoInput): Promise<foo.bar.ProtoResult> {
+  return foo.bar.ProtoResult({ sentence: `Name is ${input.name}` });
+}
+```
+
+We originally were thinking of this, but the namespaces in `json-module.js` get lost through `patchRoot()`:
+
+```
+import * as generatedRoot from '../protos/json-module';
+
+const patchRoot = <T>(x: T): T => x;
+const root = patchRoot(generatedRoot);
+
+function myWorkflowError(input: root.foo.bar.ProtoActivityInput) {
+  return input.name;
+}
+```
+
+On root in `root.foo.bar.ProtoActivityInput`, TS errors: `Cannot find namespace 'root'.`
+
+## Future work
+
+If we can get changes merged into `protobufjs` (or want to fork), we can do one or both of the below:
+
+1. Change the `json-module` output to not have `nested` attributes so we don't have to patch
```

Comment:
Make `pbts` generate private (or protected) constructors so users don't accidentally try to use them and have them rely on the static `create` method.